concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/LFMBTree.hs

```diff
@@ -284,28 +341,85 @@ fromAscListMaybes l = fromList $ go l 0
 
 -- | Get the hash of an LFMBTree constructed from a 'Foldable' by inserting each item sequentially
 -- from index 0.
--- prop> hashFromFoldable l == getHash (fromFoldable @Word64 l)
+-- prop> hashP1FromFoldable l == getHash (fromFoldable @Word64 l)
 --
 -- TODO: Optimise this implementation.
-hashFromFoldable :: (Foldable f, HashableTo H.Hash v) => f v -> H.Hash
-hashFromFoldable = getHash . fromFoldable @Word64
+hashP1FromFoldable :: (Foldable f, HashableTo H.Hash v) => f v -> LFMBTreeHashV0
+hashP1FromFoldable = getHash . fromFoldable @Word64
 
 -- | Hash a list of hashes in the LFMBTree format, using the specified hash for the empty tree.
 --  This avoids building the full tree.
-hashAsLFMBT ::
+--  This uses the V0 hashing scheme, where the hash of a node is the hash of the concatenated
+--  hashes of its children.
+hashAsLFMBTV0 ::
     -- | Hash to use for empty list
     H.Hash ->
     -- | List of hashes to construct into Merkle tree
     [H.Hash] ->
     H.Hash
-hashAsLFMBT e = go
+hashAsLFMBTV0 e = go
   where
     go [] = e
     go [x] = x
     go xs = go (f xs)
     f (x : y : xs) = H.hashOfHashes x y : f xs
     f other = other
 
+-- | Get the hash of an LFMBTree constructed from a 'Foldable'.
+--
+-- prop> lfmbtV0Hash l == getHash (fromFoldable @Word64 l)
+lfmbtV0Hash :: (Foldable f, HashableTo H.Hash v) => f v -> LFMBTreeHashV0
+{-# INLINE lfmbtV0Hash #-}
+lfmbtV0Hash = lfmbtV0Hash' getHash
+
+-- | Get the hash of an LFMBTree constructed from a 'Foldable' by applying the given hashing
+--  function to each element.
+lfmbtV0Hash' :: (Foldable f) => (v -> H.Hash) -> f v -> LFMBTreeHashV0
+{-# INLINE lfmbtV0Hash' #-}
+lfmbtV0Hash' hsh = LFMBTreeHash . hashAsLFMBTV0 (theLFMBTreeHash emptyTreeHash) . map hsh . toList
+
+-- | Hash a list of hashes in the LFMBTree format, using the specified hash for the empty tree.
+--  This avoids building the full tree.
+--  This uses the V1 hashing scheme, where the top level hash is the hash of the concatenation of
+--  the number of leaves in the tree and the V0 hash.
+hashAsLFMBTV1 ::
+    -- | Hash to use for empty list
+    H.Hash ->
+    -- | List of hashes to construct into Merkle tree
+    [H.Hash] ->
+    H.Hash
+hashAsLFMBTV1 e l = H.hashLazy $! S.runPutLazy $ do
+    S.putWord64be (fromIntegral $ length l)
+    S.put $ hashAsLFMBTV0 e l
```

Comment:
It is indeed redundant to include the size in each subtree. The only reason I can think of that it might be useful would be in decomposing a Merkle proof for an LFMBTree, but even then I am not sure it would add any benefit. The downside would be that Merkle proofs would become longer, and probably more complex to check and construct.

I have added a comment elsewhere explaining the benefit of having the size at the root.